src/io/scratchBLE.js

```diff
@@ -0,0 +1,104 @@
+const JSONRPCWebSocket = require('../util/jsonrpc-web-socket');
+const PeripheralChooser = require('./peripheralChooser');
+
+const ScratchLinkWebSocket = 'ws://localhost:20110/scratch/ble';
+
+class ScratchBLE extends JSONRPCWebSocket {
+    constructor () {
+        const ws = new WebSocket(ScratchLinkWebSocket);
+
+        super(ws);
+
+        this._ws = ws;
+        this.peripheralChooser = new PeripheralChooser(); // TODO: finalize gui connection
+        this._characteristicDidChange = null;
+    }
+
+    /**
+     * Returns a promise for when the web socket opens.
+     * @return {Promise} - a promise when BLE socket is open.
+     */
+    waitForSocket () {
+        return new Promise((resolve, reject) => {
+            this._ws.onopen = resolve;
+            this._ws.onerror = reject;
+        });
+    }
+
+    /**
+     * Request a device with the device options and optional gui options.
+     * @param {object} deviceOptions - list of device guiOptions.
+     * @param {object} onConnect - on connect callback.
+     * @param {object} onError - on error callbackk.
+     */
+    requestDevice (deviceOptions, onConnect, onError) {
```

Comment:
This (or someone else, the peripheral chooser?) should probably handle what happens when 'discover' doesn't find any devices after a while